sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -175,6 +175,8 @@ object Cast {
 
     case (from: UserDefinedType[_], to: UserDefinedType[_]) if to.acceptsType(from) => true
 
+    case (_: YearMonthIntervalType, _: YearMonthIntervalType) => true
```

Comment:
Could you open a separate PR for this, and write test only casts. It seems we need the same for the `DayTimeIntervalType`. Please, do that in the same PR.

Reply:
> Could you open a separate PR for this, and write test only casts. It seems we need the same for the `DayTimeIntervalType`. Please, do that in the same PR.

Done in https://github.com/apache/spark/pull/33035

Reply:
@AngersZhuuuu Thank you.